### PR TITLE
Paired move truncation

### PR DESCRIPTION
**Pull request by Geotale**
Matches hardware in rounding paired singles after move operations!!
Move operations consist of operations which only transfer direct bits (including absolute values and merges)
This also accounts for `ps_rsqrte` because it has a similar quirk
Specifically in hardware they're rounded in accordance to their slot:
- PS0 rounds *only the mantissa* in accordance to the set rounding mode
- PS1 truncates the mantissa

`ps_rsqrte` also only truncates for PS0 for some reason ^^;
This has all been tested on hardware, along with a few edge case tests to ensure this implementation works

### CODE REVIEW

**Comment by Dentomologist** [comment]
Can you add a unit test matching the tests you ran on hardware? That would help show what has been tested and prevent any future regressions.

Also, the second commit should be squashed into the first one.

**Comment by Geotale** [comment]
I can probably commit to the hwtests repo if I can figure out how to get it set up ^^;
I also will definitely squash the commits but I am not confident that I won't have to change things and I'd rather not have to multiple times if possible :D

**Comment by Dentomologist** [comment]
While it would definitely be good to have that in the hwtests repo, I was primarily thinking about [here](https://github.com/dolphin-emu/dolphin/tree/master/Source/UnitTests/Core/PowerPC) which gets run on every PR.

**Comment by Geotale** [comment]
I'm mildly afraid to try that bc this is interpreter-only and would require an entire thing somehow setting that up :,D

**Comment by JosJuice** [comment]
I would very much like a hwtest. And if we have that, I don't think there's any need to require a unit test on top. Yes, being able to automatically test stuff is nice, but:

1. This code doesn't get changed often at all, so running the relevant hwtests manually is a surmountable task, and
2. The time it would take to set up a bunch of unit tests for the interpreter would probably be better spent getting hwtests to run automatically like we run FifoCI, or setting up some other kind of testing system that can run both on console and on Dolphin. Unit tests are inherently less useful than hwtests because we can't run them on hardware.

File: Source/Core/Core/PowerPC/Interpreter/Interpreter_Paired.cpp [hunk 1]
```diff
@@ -10,6 +10,66 @@
 #include "Core/PowerPC/Interpreter/Interpreter_FPUtils.h"
 #include "Core/PowerPC/PowerPC.h"
 
+// Instructions which move data without performing operations round a bit weirdly
+// Specifically they rounding the mantissa to be like that of a 32-bit float,
+// going as far as to focus on the rounding mode, but never actually care about
+// making sure the exponent becomes 32-bit
+// Either this, or they'll truncate the mantissa down, which will always happen to
+// PS1 OR PS0 in ps_rsqrte
+inline u64 TruncMantissaBits(u64 bits)
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/Interpreter/Interpreter_Paired.cpp:19]:
```suggestion
inline u64 TruncateMantissaBits(u64 bits)
```

File: Source/Core/Core/PowerPC/Interpreter/Interpreter_Paired.cpp [hunk 2]
```diff
@@ -10,6 +10,66 @@
...
+{
+  // Truncation can be done by simply cutting off the mantissa bits that don't
+  // exist in a single precision float
+  u64 remove_bits = Common::DOUBLE_FRAC_WIDTH - Common::FLOAT_FRAC_WIDTH;
+  u64 remove_mask = (1 << remove_bits) - 1;
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/Interpreter/Interpreter_Paired.cpp:24]:
```suggestion
  constexpr u64 remove_bits = Common::DOUBLE_FRAC_WIDTH - Common::FLOAT_FRAC_WIDTH;
  constexpr u64 remove_mask = (1 << remove_bits) - 1;
```

File: Source/Core/Core/PowerPC/Interpreter/Interpreter_Paired.cpp [hunk 3]
```diff
@@ -10,6 +10,66 @@
...
+  return bits & ~remove_mask;
+}
+
+inline double TruncMantissa(double value)
```
Comment (JosJuice) [inline Source/Core/Core/PowerPC/Interpreter/Interpreter_Paired.cpp:28]:
```suggestion
inline double TruncateMantissa(double value)
```

**Comment by Tilka** [comment]
I've confirmed that the ps_neg/ps_abs/ps_nabs changes match hardware. If you move the changes for those opcodes (and the shared code) into a separate PR, we can get that merged now and leave ps_sel/ps_rsqrte/ps_sum0/ps_sum1 verification for later.

**Comment by Geotale** [comment]
Oh right I forgot about sum0 and sum1, ty for reminding me ^^ -- I was planning try to figure out what in the world is failing the 3 specific tests with ps_rsqrte first though :O

**Comment by Tilka** [comment]
Based on some limited testing, I think ps_sel doesn't have the NaN/infinity special case.

**Comment by Geotale** [comment]
I did get the tests fully working, but the changes might be larger and more complicated than initially expected ^^; I'm going to abstract as much away as possible though

**Comment by Geotale** [comment]
ARM tests are actually failing which I think is really funny ^^; I would really like more information to see what exactly is failing and can try to fix the JIT (or my own code!!) to resolve the issue need be!!

**Comment by AdmiralCurtiss** [comment]
You can see it in the build log for the failing build:

```
[ RUN      ] JitArm64.Fres
3ff0000000000000 -> 3ffffe0000000000 == 3fefff0000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 4607181319288389632
  actual
    Which is: 4611683819404132352
bff0000000000000 -> bffffe0000000000 == bfefff0000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 13830553356143165440
  actual
    Which is: 13835055856258908160
3800000000000000 -> 47dffe0000000000 == 47dfff0000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 5179138471964442624
  actual
    Which is: 5179137372452814848
3810000000000000 -> 47dffe0000000000 == 47cfff0000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 5174634872337072128
  actual
    Which is: 5179137372452814848
b800000000000000 -> c7dffe0000000000 == c7dfff0000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 14402510508819218432
  actual
    Which is: 14402509409307590656
b810000000000000 -> c7dffe0000000000 == c7cfff0000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 14398006909191847936
  actual
    Which is: 14402509409307590656
3800123456789abc -> 57f0212700000000 == 47dfdbd8e0000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 5179099821016875008
  actual
    Which is: 6336601127097729024
3810123456789abc -> 57e0212700000000 == 47cfdbd8e0000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 5174596221389504512
  actual
    Which is: 6332097527470358528
b800123456789abc -> d7f0212700000000 == c7dfdbd8e0000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 14402471857871650816
  actual
    Which is: 15559973163952504832
b810123456789abc -> d7e0212700000000 == c7cfdbd8e0000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 14397968258244280320
  actual
    Which is: 15555469564325134336
47c0000000000000 -> 381ffe0000000000 == 381fff0000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 4044231365867077632
  actual
    Which is: 4044230266355449856
47d0000000000000 -> 0000000000000000 == 380fff0000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 4039727766239707136
  actual
    Which is: 0
c7c0000000000000 -> b81ffe0000000000 == b81fff0000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 13267603402721853440
  actual
    Which is: 13267602303210225664
c7d0000000000000 -> 8000000000000000 == b80fff0000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 13263099803094482944
  actual
    Which is: 9223372036854775808
37f0000000000000 -> 47fffe0000000000 == 47efff0000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 5183642071591813120
  actual
    Which is: 5188144571707555840
b7f0000000000000 -> c7fffe0000000000 == c7efff0000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 14407014108446588928
  actual
    Which is: 14411516608562331648
3ff8000000000000 -> 3feaaa0000000000 == 3fe5550000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 4604179652544561152
  actual
    Which is: 4605680485916475392
408f400000000000 -> 3f51280000000000 == 3f50628000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 4562254724421648384
  actual
    Which is: 4562471877968134144
c008000000000000 -> bfdaaa0000000000 == bfd5550000000000
/Users/administrator/buildbot-worker/osx_m1/pr-osx-universal/build/Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:72: Failure
Expected equality of these values:
  expected
    Which is: 13823048089771966464
  actual
    Which is: 13824548923143880704
[  FAILED  ] JitArm64.Fres (1 ms)
```

**Comment by Geotale** [comment]
I can check on hardware quickly, but at least from the `3ff0000000000000` -> `3fefff0000000000` (new result) it seems like this PR is the correct one?
Although I'll also add that some of these are issues with denormals -- I'll update the tested FPSCR if we should just always assume that the NI flag is set ^^

**Comment by Geotale** [comment]
Thinking about it, actually I do need to modify the JIT if it ends up calling this function!! x86-64 definitely does this at least...

**Comment by Geotale** [comment]
I'm going to try my best as of now for x86-64 at the very least because the outputs of that function seem to match those failing in ARM64 on my machine :o It's so weird too...

**Comment by Geotale** [comment]
Although the build should pass, note that this PR can't really be submitted yet until it's decided for JITs in the `fres` instruction whether:
a. x86-64 should have imperfect behavior and inline more for more complex inputs, or
b. Aarch64 should instead use a function call for complex inputs, giving perfect behavior
Specifically the inputs differ for numbers which would give a subnormal output when NI=0 (but never in any other instances)
Also technically a function call isn't necessary for either of them, because both implement the most complex part, and Aarch64 implements everything other than this subnormal case ^^

File: Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp [hunk 4]
```diff
@@ -40,6 +40,7 @@ class TestFres : public JitArm64
     MOV(ARM64Reg::X1, ARM64Reg::X0);
     m_float_emit.FMOV(ARM64Reg::D0, ARM64Reg::X0);
     m_float_emit.FRECPE(ARM64Reg::D0, ARM64Reg::D0);
+    m_float_emit.FMOV(ARM64Reg::X0, ARM64Reg::D0);
```
Comment (JosJuice) [inline Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:43]: What's the purpose of this? The fres routine already does `m_float_emit.FMOV(ARM64Reg::X0, ARM64Reg::D0)`.

File: Source/Core/Common/FloatUtils.h [hunk 5]
```diff
@@ -96,6 +97,88 @@ extern const std::array<BaseAndDec, 32> fres_expected;
...
+  u64 bits = std::bit_cast<u64>(value);
+  u64 trunc_bits = TruncateMantissaBits(bits);
+  return std::bit_cast<double>(trunc_bits);
+}
+
+inline u64 RoundMantissaBitsFinite(u64 bits)
+{
+  const u64 replacement_exp = 0x4000000000000000ull;
+
+  // To round only the mantissa, we assume the CPU can change the rounding mode,
```
[the remark this answers is not in the document]
Comment (Geotale) [reply Source/Core/Common/FloatUtils.h:128]: It doesn't change the rounding mode, but it assumes the CPU has properly changed the rounding mode to match PowerPC's, so rounding from a double to a float goes in the right direction!

File: Source/Core/Common/FloatUtils.h [hunk 6]
```diff
@@ -96,6 +97,88 @@ extern const std::array<BaseAndDec, 32> fres_expected;
...
+  // for some reason it assumes that ps0 should be rounded with
+  // finite values rather than checking if they might be infinite
+  u64 bits = std::bit_cast<u64>(value);
+  u64 rounded_bits = RoundMantissaBitsFinite(bits);
+  return std::bit_cast<double>(rounded_bits);
+}
+
+inline double RoundMantissa(double value)
+{
+  // The double version of the function just converts to and from bits again
```
[the remark this answers is not in the document]
Comment (Geotale) [reply Source/Core/Common/FloatUtils.h:177]: I meant versus the version that takes a u64 input as bits, but I agree that the comment should be changed to clear that up

File: Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp (hunk 4 above)
Comment (Geotale) [reply Source/UnitTests/Core/PowerPC/JitArm64/Fres.cpp:43]: I didn't consider it being part of the FRES routine already, thank you!!

File: Source/Core/Core/PowerPC/Interpreter/Interpreter_Paired.cpp [hunk 7]
```diff
@@ -171,7 +181,14 @@ void Interpreter::ps_rsqrte(Interpreter& interpreter, UGeckoInstruction inst)
 {
   auto& ppc_state = interpreter.m_ppc_state;
   const double ps0 = ppc_state.ps[inst.FB].PS0AsDouble();
-  const double ps1 = ppc_state.ps[inst.FB].PS1AsDouble();
+  double ps1 = ppc_state.ps[inst.FB].PS1AsReciprocalDouble();
+
+  if (ps1 > 0.0)
```
[the remark this answers is not in the document]
Comment (Geotale) [reply Source/Core/Core/PowerPC/Interpreter/Interpreter_Paired.cpp:186]: It's not, no, I didn't know CPUs did that :O A simple bitwise comparison should be fine instead though

File: Source/Core/Common/FloatUtils.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (JosJuice) [reply Source/Core/Common/FloatUtils.h:128]: It's fine to rely on the host CPU's rounding mode being set to match the emulated CPU. The comment would be clearer if changed to something like "we assume the CPU's rounding mode is correctly set".

File: Source/Core/Core/PowerPC/Interpreter/Interpreter_Paired.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (JosJuice) [reply Source/Core/Core/PowerPC/Interpreter/Interpreter_Paired.cpp:186]: I think I was a little too hasty when writing this comment! So, on PowerPC, there's the NI bit, which will make outputs be flushed to zero if they're subnormals. This does *not* flush inputs to zero when they're subnormals. On x86, you can separately set whether to flush inputs and flush outputs, so we set host x86 CPUs to match the PowerPC behavior. On AArch64, it gets trickier. There's an extension on recent CPUs that lets you control the two separately like on x86, but Dolphin still runs on a lot of CPUs that don't have that extension. On those older AArch64 CPUs, when FPSCR.NI is set, we configure the host CPU to flush both inputs and outputs. From testing, this seems to make games behave more closely to real consoles than not flushing at all.

So on certain AArch64 CPUs only, inputs will be flushed, i.e. subnormals will be treated as if they're equal to 0 for comparison operators like these. If you could do a bitcast to u64 and do the comparison on those bits instead, that would fix the issue. But if by conversion you mean converting double to float, that operation is also affected. (Though, now that I think about it, that's not relevant anyway... Converting double to float always causes subnormals to become zero, due to the limited precision of floats...)

File: Source/Core/Common/FloatUtils.cpp [hunk 8]
```diff
@@ -2,6 +2,7 @@
 // SPDX-License-Identifier: GPL-2.0-or-later
 
 #include "Common/FloatUtils.h"
+#include "Core/PowerPC/Gekko.h"
```
[the remark this answers is not in the document]
Comment (Geotale) [reply Source/Core/Common/FloatUtils.cpp]: Moved to Core ^^b It compiles and I've checked over all instances of the file so this should be resolved. I opted to move it to Core for very very minor optimization purposes and the reasons you recommended

File: Source/Core/Common/FloatUtils.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (Geotale) [reply Source/Core/Common/FloatUtils.h]: Changed to instead say that we assume the host CPU's rounding mode matches that of the emulated CPU

File: Source/Core/Core/PowerPC/Interpreter/Interpreter_Paired.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Geotale) [reply Source/Core/Core/PowerPC/Interpreter/Interpreter_Paired.cpp]: This should be resolved now!! Checks if the bits are signed greater than 0

**Comment by Geotale** [comment]
Git literally why
